Multigateway support for Egress Gateway

carlos-abad opened the pull request:
Add Multigateway support to egress gateway. 

[Link to the CFP](https://github.com/cilium/design-cfps/blob/main/cilium/CFP-38341-multigateway-egress-gateway-policy.md)

Fixes: #38341

```release-note
Add support for multiple gateways to Cilium Egress Gateway.
```

carlos-abad committed:
pkg/egressgateway: add multigateway to CEGP API

Signed-off-by: Carlos Abad <carlosab@google.com>

carlos-abad committed:
pkg/egressgateway: move IPv6 indicator to PolicyConfig

Since the boolean v6needed is determined from the CEGP destinationCIDRs
field, we can move it to the PolicyConfig object rather than the
PolicyGatewayConfig.

This change will be very useful for the multigateway Egress Policy
since we plan to have multiple policyGatewayConfig objects per
PolicyConfig. This change will safe from storing the exact same boolean
value in all policyGatewayConfig objects.

Signed-off-by: Carlos Abad <carlosab@google.com>

carlos-abad committed:
pkg/egressgateway: Add support for parsing multigateway CEGP

This change adds support for parsing the multiple gateways
from a multigateway CEGP and storing them in a PolicyConfig object.

This change doesn't modify how the BPF tables are written.

Signed-off-by: Carlos Abad <carlosab@google.com>

carlos-abad committed:
pkg/egressgateway: Add multigateway assignment

As discussed in the design the assingment follows a round-robin
algorithm. To create repeatable and stable assignments the lists of
gaterways and endpoints are ordered before the rounds robin-assignment.

Design: CFP-38341-multigateway-egress-gateway-policy

Signed-off-by: Carlos Abad <carlosab@google.com>

carlos-abad committed:
pkg/egressgateway: Add documentation for multigateway

Signed-off-by: Carlos Abad <carlosab@google.com>

tommyp1ckles commented on pkg/k8s/apis/cilium.io/v2/cegp_types.go:
```diff
@@ -63,7 +63,17 @@ type CiliumEgressGatewayPolicySpec struct {
 	EgressGateway *EgressGateway `json:"egressGateway"`
+
+	// Optional list of gateway nodes responsible for SNATing traffic.
+	// The first entry must contain the gateway node from the EgressGateway field.
```
This seems confusing to me, why not just have the EgressGateways field take precedence over the old singular one.  Such that if EgressGateways is not nil then we simply ignore EgressGateway.

EgressGateways just seems like a generalization of EgressGateways anyway so in my mind I would expect have EgressGateways of length one to be equivalent to using EgressGateway (assuming they're the same egress-gateway).

carlos-abad replied:
Sounds good. I'll change it to behave that way.

tommyp1ckles commented on pkg/egressgateway/policy.go:
```diff
@@ -68,6 +67,8 @@ type PolicyConfig struct {
 	matchedEndpoints map[endpointID]*endpointMetadata
 	gatewayConfig    gatewayConfig
+
+	v6needed bool
```
nit: `v6Needed` is more idiomatic

carlos-abad replied:
Done.

qmonnet commented on examples/kubernetes-egress-gateway/egress-gateway-policy.yaml:
```diff
@@ -34,3 +34,24 @@ spec:
     # In this case the first IPv4 assigned to the interface with the default
     # route will be used as egressIP
+  egressGateways:
+    # It's possible to specify multiple egres gateways. In this case the source
```
```suggestion
    # It's possible to specify multiple egress gateways. In this case the source
```

carlos-abad replied:
Done

qmonnet commented on examples/kubernetes-egress-gateway/egress-gateway-policy.yaml:
```diff
@@ -34,3 +34,24 @@ spec:
+    # It's possible to specify multiple egres gateways. In this case the source
+    # endpoints will egress traffic through one of the gateways listed below.
+    # 
+    # The configuration of each gatewayis the same as the field "egressGateway"
```
```suggestion
    # The configuration of each gateway is the same as the field "egressGateway"
```

carlos-abad replied:
Done.

qmonnet commented on examples/kubernetes-egress-gateway/egress-gateway-policy.yaml:
```diff
@@ -34,3 +34,24 @@ spec:
+    # The configuration of each gatewayis the same as the field "egressGateway"
+    # above.
+    #
+    # The first element has to be the same as the egressGateway above. 
```
I first had the impression that the two sentences meant the same thing. Maybe expend a little bit. I found your explanation in the docs clearer.

carlos-abad replied:
Sounds good. I've reworded them to be closer to the documentation.

qmonnet reviewed (changes requested):
Some nits on the docs.

Regarding the UX, I agree with Tom: do we really need this requirement on the first item of the list being in `egressGateway` as well?

qmonnet reviewed (approved):
Looks OK from my side, thank you!

aanm commented on pkg/k8s/apis/cilium.io/v2/cegp_types.go:
```diff
@@ -63,7 +63,17 @@ type CiliumEgressGatewayPolicySpec struct {
 	ExcludedCIDRs []CIDR `json:"excludedCIDRs"`
 
 	// EgressGateway is the gateway node responsible for SNATing traffic.
+	// In case multiple nodes are a match for the given set of labels, the first node
```
Because of this change, we need to bump the schema version defined in pkg/k8s/apis/cilium.io/register.go

carlos-abad replied:
Done.

tommyp1ckles commented on pkg/egressgateway/policy.go:
```diff
@@ -233,18 +236,33 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(logger *slog.Logger, gc
+
+	for i, endpoint := range endpoints {
+		// Iterate over the list of gateways in a round-robin fashion.
+		gateway := &config.gatewayConfigs[i%len(config.gatewayConfigs)]
```
My main concern here is that this selection approach on its own seems quite unstable.  `forEachEndpointAndCIDR(...)` is used to write out the final policy to the egress-policy bpf map, so it may be run each time there is a relevant event (i.e. policy update/delete, endpoint create/delete, etc...).

Specifically, if a endpoint is deleted that happens to be first in the lexicographic sort (i.e. by its unique ID) then deleting that endpoint may cause all other endpoints to be assigned new gateways. For example:

Say we have endpoints: ep0, ep1, ep2 and a policy selecting gateways: gw0, gw1, gw2.  During our first egress-gateway reconciliation we would expect the following egress-policy gateway assignments:

ep0 → gw0
ep1 → gw1
ep2 → gw2

Now say ep0 is deleted, this causes the egress-gateway manager to reconcile, performing the round-robin assignment:

ep1 → gw0
ep2 → gw1

In this case, all remaining endpoints have had their policies updated with a new gateway.  Although we've left the problem of the connection black holing to future work - I think this particular problem is still worth addressing now.  That is, if we can make the gateway assignment "sticky" during subsequent reconciliations that would have several benefits:

* While we wait for a solution to the "black-hole" problem, this would minimize the chances of that happening.  Otherwise the initial release of this feature will be much more prone to broken connections.
* Possible/proposed solutions to the black-hole problem may not be without some cost, so it may be worth avoiding these changes if possible.
* From a future debugging perspective, it would be helpful to know that a endpoints assigned egress gateway will not change _unless_ that gateway has been deleted (or similar).

tommyp1ckles replied:
To accomplish this behaviour, we could probably do a bpf map lookup on the policy key in `addEgressRule` and check if the policy already exists and if so, check if the assigned gateway is still a valid gateway:

* If a policy exists, and gateway is valid, then use the old gateway.
* Otherwise, if the gateway is not valid, use the one selected by `gateway := &config.gatewayConfigs[i%len(config.gatewayConfigs)]`.

carlos-abad replied:
> we could probably do a bpf map lookup on the policy key in addEgressRule and check if the policy already exists 

As I mentioned in the CFP PR, that will cause different Cilium nodes to generate different egressgateway BPF tables and break the functionality. If we apply this solution to the example you mentioned above, where ep0 is deleted, if a new Cilium node comes online, that node won't have any entries in its BPF table. Then the nodes without the BPF table will assign a different gateway to the endpoint 1 (gw0) than the nodes that had that an entry in the BFP table (gw1):

Node 1 (new node, BPF table clean)       
ep1 → gw0                                       
ep2 → gw1                                       

Node 2 (didn't restart, had info in the BPF table)
ep1 → gw1
ep2 → gw2

carlos-abad replied:
> In this case, all remaining endpoints have had their policies updated with a new gateway. Although we've left the problem of the connection black holing to future work - I think this particular problem is still worth addressing now. That is, if we can make the gateway assignment "sticky" during subsequent reconciliations

Given the issue that I brought up in my previous comment, it's not possible to have a fully sticky assignment without sharing the assignment information between the nodes. This is tricky because it requires coordination between the nodes. As such, I want to leave solving that issue outside of the scope of this PR. 

One thing I can do in the scope of this PR is to do the gateway assignment based on the hash of the ID of the endpoint, so the same endpoint gets consistently assigned to the same gateway.

`gateway := &config.gatewayConfigs[hash(endpoint.id)%len(config.gatewayConfigs)]`

Let me know what you think and if so I'll modify the PR with that change.

tommyp1ckles replied:
> Given the issue that I brought up in my previous comment, it's not possible to have a fully sticky assignment without sharing the assignment information between the nodes. This is tricky because it requires coordination between the nodes. As such, I want to leave solving that issue outside of the scope of this PR.

Ah sorry - missed the last comment.  

Per the comment in the CFP:

>  In that case, the new node won't know that B was assigned to gw2, so that alternative won't work.

Not sure I follow here, an endpoint is matched 1-1 with a (cilium managed) Pod resource which can't be moved across Nodes.

> Node 1 (new node, BPF table clean)
> ep1 → gw0
> ep2 → gw1
> 
> Node 2 (didn't restart, had info in the BPF table)
> ep1 → gw1
> ep2 → gw2

So I guess I'm confused by the example here, as Node2 would never have ep1/ep2 running on it.

tommyp1ckles replied:
One thing I think I missed with this was, that we currently watch/create policies for all endpoints, rather than just local ones.  I think this is to have the policy be in place such that all nodes can become gateways.  Would having inconsistent gateway fields in realized egress policy across nodes be an issue (cc @julianwiedmann @ysksuzuki)?

carlos-abad replied:
> So I guess I'm confused by the example here, as Node2 would never have ep1/ep2 running on it.

Let me step back for a second. When you create a CEGP in a cluster all the Cilium nodes in the cluster add the entries for that CEGP to their egress gateway BPF table. Even the nodes that don't have those endpoint. This include the gateway nodes. 

Just to double check I did an experiment in a kind cluster with a simple CEGP. All the non-gateway nodes got the same entry in the egress gateway BFP table:
```
kubectl -n kube-system exec pod/cilium-jskcm -- cilium-dbg bpf egress list
Source IP     Destination CIDR   Egress IP   Gateway IP
10.244.2.17   0.0.0.0/0          0.0.0.0     192.168.10.4

kubectl -n kube-system exec pod/cilium-mtrv2 -- cilium-dbg bpf egress list
Source IP     Destination CIDR   Egress IP   Gateway IP
10.244.2.17   0.0.0.0/0          0.0.0.0     192.168.10.4
```

The gateway node added an entry including the egress IP as expected:
```
kubectl -n kube-system exec pod/cilium-4p58m -- cilium-dbg bpf egress list
Source IP     Destination CIDR   Egress IP       Gateway IP
10.244.2.17   0.0.0.0/0          10.168.60.100   192.168.10.4
```

As you can see, the source node and the gateway node must have coordinated entries. Otherwise the gateway node won't know what to do with the packet that it receives from the source node.

Now let's start again with the example from before.

The source node has initially this configuration:
ep0 → gw0
ep1 → gw1
ep2 → gw2

And the gateway node too:
ep0 → gw0
ep1 → gw1
ep2 → gw2

Now imagine that and endpoint (ep0) is deleted and the gateway node (gw1) fully restarts. In that case, with your proposed solution, the source node will use the egress gateway BPF map to recreate the assignments, but the gateway can't because it just restarted. In this case maps will end up as:

The source node:
ep1 → gw1
ep2 → gw2

Gateway node:
ep1 → gw0
ep2 → gw1

In this case when a packet from the source node (ep1) gets to the gateway (gw1), that gateway won't know it should act as the gateway for that endpoint. 

Hence, we need to be able to recreate the egress gateway bpf map consistently in all nodes without using any previous state on the node.

tommyp1ckles replied:
Ok, I understand now.  I was missing that we need the gateway to be consistent in order for [gateways to know ](https://github.com/cilium/cilium/blob/main/bpf/lib/egress_gateway.h#L257)when they should snat the traffic.

So, going back to my rationale for all this, I do think that having unstable selection will cause this functionality to be very difficult to use as each endpoint update would cause a new/different policy to be written.

> One thing I can do in the scope of this PR is to do the gateway assignment based on the hash of the ID of the endpoint, so the same endpoint gets consistently assigned to the same gateway.
>
> gateway := &config.gatewayConfigs[hash(endpoint.id)%len(config.gatewayConfigs)]

This seems like it would improve this, as it would prevent re-computation except in the case that config.gatewayConfigs changes.

carlos-abad replied:
Sounds good. I modified the PR to use the hash method instead. 

As we discussed over chat this change should address this concern, so I'll resolve this conversation.

carlos-abad committed:
pkg/egressgateway: Mutigateway E2E connectivity test.

To run this test manually please set the kind cluster using the egressgw
preset:
- make kind-egressgw
- make kind-image
- make kind-egressgw-install-cilium

And run it in unsafe mode.
cilium connectivity test --include-unsafe-tests --test seq-egress-gateway-multigateway

Signed-off-by: Carlos Abad <carlosab@google.com>

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+		for _, echo := range ct.EchoPods() {
+			assignedGateway := getGatewayForIP(echo.Pod.Status.PodIP)
+			if assignedGateway == nil {
+				t.Fatalf("Couln't find gateway for the client pod %s/%s with IP: %s",
```
typo Couln't -> Couldn't

carlos-abad replied:
Fixed it.

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+		panic("no gateway nodes available for assignment")
+	}
+	h := fnv.New32a()
+	h.Write([]byte(epUID))
```
nit: little bit of overlap with the business logic over in the egress manager, might be worth making a note of that.

carlos-abad replied:
Added a note

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+		for _, client := range ct.ClientPods() {
+			assignedGateway := getGatewayForIP(client.Pod.Status.PodIP)
+			if assignedGateway == nil {
+				t.Fatalf("Couln't find gateway for the client pod %s/%s with IP: %s",
```
typo Couln't -> Couldn't

carlos-abad replied:
Fixed it.

tommyp1ckles commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -303,6 +303,40 @@ setting the ``--devices`` agent option accordingly.
+It's possible to select multiple gateway nodes in the same policy. In this case, the gateway nodes
+can be configured using the ``egressGateways`` list field. Entries on this list have the exact same
+configuration options as the ``egressGateway`` field:
+
```
I think it would be helpful for users to have more information regarding this functionality and the implications of using such as:

* Some details of the gateway selection algorithm (i.e. that endpoints will be assigned a gateway based on their unique CEP ID).
* Caveats and risks of using multiple gateways.

carlos-abad replied:
Added some notes and a warning section with that information.

tommyp1ckles commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -303,6 +303,40 @@ setting the ``--devices`` agent option accordingly.
+
+.. code-block:: yaml
+
+  egressGateway:
```
Leaving this field in the example makes it unclear if it's necessary for using the multiple gateways functionality.

carlos-abad replied:
Good point. I removed that field to avoid confusion.

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+		})
+	}
+
+	// Sort gateway nodes by their IPv4 internal IP to ensure deterministic assignment
```
Question, aren't these already sorted by the egress manager on the agent: https://github.com/cilium/cilium/blob/d3b0bc33989bac27c22fefb108e7c7dce5039397/pkg/egressgateway/policy.go#L249-L251.  ~In which case, do we need to re-sort?~

tommyp1ckles replied:
Oh I see, this is because `assignGatewayNode` expects a sorted list to duplicate the agents selection logic. Perhaps we should mention that in a comment (or just sort inside assignGatewayNode? no real peformance concerns here).

carlos-abad replied:
I added some comments clarifying it.

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+
+func assignGatewayNode(epUID string, sortedGatewayNodes []gatewayNodeInfo) *gatewayNodeInfo {
+	if len(sortedGatewayNodes) == 0 {
+		panic("no gateway nodes available for assignment")
```
Can we just let it return nil and have the test fail due to the `assignedGateway == nil check` instead of calling panic.  Panic in the connectivity tests should be only in the worst case it prevents the normal cilium-cli connectivity test post-test cleanups from happening.

carlos-abad replied:
SG. I'll do that.

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+	err = check.WaitForEgressGatewayBpfPolicyEntries(ctx, ct.CiliumPods(), func(ciliumPod check.Pod) ([]check.BPFEgressGatewayPolicyEntry, error) {
+		var targetEntries []check.BPFEgressGatewayPolicyEntry
+
+		for _, client := range ct.ClientPods() {
```
Can we pass `ct.ClientPods()` to `WaitForEgressGatewayBpfPolicyEntries` instead of `ct.CiliumPods()` to avoid this inner loop?

carlos-abad replied:
I don't think I understand, because passing `ct.ClientPods()` won't remove the need for the inner loop. We still need to create the expected BPF table and for that we need to loop over `ct.ClientPods()`. Also, egressGateway.Run() also has a similar loop.

IIUC, the only effect of passing `ct.ClientPods()` to `WaitForEgressGatewayBpfPolicyEntries()` would be that we only check the BPF table in the client pods and not all the pods.

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+
+	// Ping hosts (pod to host connectivity). Should not get masqueraded with egress IP
+	i := 0
+	for _, client := range ct.ClientPods() {
```
nit: this check is identical to existing tests, we can probably just put this into a common function to reduce code duplication.

carlos-abad replied:
Good point. I'll extract some common code into functions.

tommyp1ckles commented on cilium-cli/connectivity/tests/egressgateway.go:
```diff
@@ -276,6 +280,341 @@ func (s *egressGateway) Run(ctx context.Context, t *check.Test) {
+	// This test case fails for ipv6, might need to make changes for this to work with
+	// kube-dns or target a different in-cluster service.
+	i = 0
+	for _, client := range ct.ClientPods() {
```
Same question here, this seems identical to other such checks for the single GW tests, we can probably DRY this up.

carlos-abad replied:
Same as the previous comment, I extracted all the common code into common functions.

tommyp1ckles commented on pkg/egressgateway/manager_privileged_test.go:
```diff
@@ -766,6 +778,182 @@ func TestEndpointDataStore(t *testing.T) {
+		var rules []egressRule
+
+		for _, endpoint := range endpoints {
+			h := fnv.New32a()
```
nit: we could just use computeEndpointHash here

carlos-abad replied:
Done.

tommyp1ckles commented on pkg/egressgateway/manager_privileged_test.go:
```diff
@@ -766,6 +778,182 @@ func TestEndpointDataStore(t *testing.T) {
+	addEndpoint(t, k.endpoints, eps[0].ep)
+	reconciliationEventsCount = waitForReconciliationRun(t, egressGatewayManager, reconciliationEventsCount)
+	assertEgressRules4(t, policyMap4, assignEndpoints(eps, nodes, true))
+	assertEgressRules6(t, policyMap6, assignEndpoints(eps, nodes, false))
```
nit: One more test assertion that would be nice to have is checking that the remaining gateways endpoint selection does not change following remove/re-adding of the endpoint (i.e. from the UID hashing).

carlos-abad replied:
Excellent idea. I added it to the test.

tommyp1ckles commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -303,6 +303,44 @@ setting the ``--devices`` agent option accordingly.
+
+.. warning::
+
+    As with single-gateway policies, changing the gateway node will cause existing egress conections
```
> blackhole traffic

Might be more in line with existing documentation to just say it will break existing egress connections

tommyp1ckles replied:
Also, typo: conections -> connections

carlos-abad replied:
Thank you for pointing it out. I updated that section.

tommyp1ckles commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -303,6 +303,44 @@ setting the ``--devices`` agent option accordingly.
+.. warning::
+
+    As with single-gateway policies, changing the gateway node will cause existing egress conections
+    to blackhole traffic. There is already a
```
> There is already Github discussion to mitigate this issue.

Can we reword this a bit, maybe just say something like:

Please see the following [Github discussion](https://github.com/cilium/cilium/discussions/39245) which tracks this issue. 

As a tangent, I've never seen a github discussion referenced in the docs, grepping through the docs pages we commonly just reference an issue. Wonder if we should turn this back to an issue to reference here (cc @joestringer thoughts?)

carlos-abad replied:
Good point. I added the link as a gh-issue rather than a "discussion" since gh-discussion is not available. However, the link redirects to the discussion anyway.

tommyp1ckles reviewed (commented):
@carlos-abad Code wise everything is looking good, just had a few more small things about the docs.

christarazi commented on cilium-cli/connectivity/check/test.go:
```diff
@@ -553,6 +556,19 @@ func (t *Test) WithCiliumEgressGatewayPolicy(params CiliumEgressGatewayPolicyPar
+		}
+
+		for _, node := range egressGatewayNodes {
+			gw := pl.Spec.EgressGateway.DeepCopy()
+			gw.NodeSelector.MatchLabels["kubernetes.io/hostname"] = node
+			pl.Spec.EgressGateways = append(pl.Spec.EgressGateways, *gw)
```
For now, is it expected that the `EgressGateways` field contains a duplicate of the `EgressGateway`? Probably worth a comment.

carlos-abad replied:
If the field `EgressGateways` is set, the contents of the field `EgressGateway` are disregarded. I'll add a clarifying comment.

derailed reviewed (approved):
@carlos-abad Nice work!

giorio94 commented on .github/actions/cl2-modules/egw/manifests/egw-policy.yaml:
```diff
@@ -16,3 +16,4 @@ spec:
     nodeSelector:
       matchLabels:
         {{StructuralData .NodeSelector}}
+  egressGateways: []
```
Why is this change necessary (and the same for similar ones)? Wouldn't requiring this modification be a breaking change of the API, that could create problems to users?

carlos-abad replied:
The problem is who this yaml files are parsed. When I leave `egressGateways` out the parsing fails because it sets that field to `null` rather than the empty array.

giorio94 replied:
Hmm, not sure I'm understanding, why are `null` and empty array being treated differently?

carlos-abad replied:
For example, if the function [newConnDisruptCEGP()](https://github.com/cilium/cilium/blob/829016fa3bcd5d7b38ce82c5c147d628759ebf0f/cilium-cli/connectivity/check/deployment.go#L587) doesn't include it, then the [ApplyGeneric()](https://github.com/cilium/cilium/blob/829016fa3bcd5d7b38ce82c5c147d628759ebf0f/cilium-cli/connectivity/check/deployment.go#L703) method fails because the new field is set as null rather than an empty array. 

You can even try it yourself if you are curious on this behaviour. Try to remove the [ExcludedCIDRs](https://github.com/cilium/cilium/blob/829016fa3bcd5d7b38ce82c5c147d628759ebf0f/cilium-cli/connectivity/check/deployment.go#L587) field from the CEGP created by `newConnDisruptCEGP()` and you'll see how you get an error during the long running e2e tests, like `Cilium E2E Upgrade`. 

I saw a similar issue with this file.

giorio94 commented on daemon/cmd/daemon_main.go:
```diff
@@ -657,6 +657,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.EnableIPv4EgressGateway)
 
+	flags.Bool(option.EnableIPv4EgressMultigateway, false, "Enable egress gateway with multiple gateways for IPv4")
+	option.BindEnv(vp, option.EnableIPv4EgressMultigateway)
```
New flags should be defined via Hive, and not included in the legacy `Config` struct. You can see an example here for instance: https://github.com/cilium/cilium/blob/8eb6dbf2fa2d74560996864c176a955c06f08385/pkg/egressgateway/manager.go#L84-L96

That said, is yet another flag really necessary? Wouldn't be the user intent of specifying `egressGateways` rather than a single `egressGateway` be enough to enable the feature? If the flag was introduced for testing purposed only, then you could simply check if the Cilium version is `>=v1.18`, as done for other features.

carlos-abad replied:
Good idea. I removed the flag and used the version check instead.

giorio94 reviewed (commented):
Thanks! A couple of comments inline for my codeowners.

pippolo84 reviewed (approved):
Thanks!

joestringer commented on examples/kubernetes-egress-gateway/egress-gateway-policy.yaml:
```diff
@@ -34,3 +34,25 @@ spec:
+    # 
+    # Entries on this list have the exact same configuration options as the
+    # EgressGateway field above.
+  - egressGateway:
+    nodeSelector:
```
Heads up that these changes were broken on arrival. Evidently we don't have CI coverage for this file, otherwise we would have noticed this earlier. But also in this case the developer didn't test this file and the reviewers didn't test this file, so we merged it in a broken state. This was fixed in https://github.com/cilium/cilium/pull/40112.

cc @cilium/egress-gateway as reviewers should we be doing anything differently to avoid introducing breakage like this into the tree?

carlos-abad replied:
I recommend we introduce automatic testing for the examples. Maybe with a separate Github flow?

By the way, just mentioning that this file already had a bug in it. When I added the new field I replicated the field above, duplicating the existing bug.

pchaigno replied:
I agree automated testing of example YAML files would be better than expect reviewers catch every indentation and whatnot issues.

joestringer replied:
Yeah improving the automated testing seems reasonable. In this case, having a short test to load the YAML using the Cilium libraries to verify that it matches the right syntax would have helped; it doesn't have to be an end-to-end test.

> By the way, just mentioning that this file already had a bug in it. When I added the new field I replicated the field above, duplicating the existing bug.

If you know about a bug, I would encourage you to submit a fix for it.

In this case the bug I was referring to is introduced by this specific change, because `EgressGateways` is defined as `[]EgressGateway` and `EgressGateway` does not have a `egressGateway` field inside of it.
